---
title: "R Notebook"
output: word_document
editor_options: 
  chunk_output_type: console
---
Database information
#Database Information
**control_expect: table of the taxonomy for expected control frequencies
--control_artcol & control_zymoext: tables of taxonomy info for artificial colony and zymo extraction only

**samples_variables_complete - table of the variable information of all samples and controls
--samples_filtered: table of all samples/controls that passed QC, filtering off of samples_failed
--samples_variables_f: subset of samples_filtered, to only include only extraction controls 

**samples_failed: table of samples that did not meet QC thresholds

**samples_otu_complete: table of the OTUs from samples_filtered by genus

**samples_variables_f: table of samples with counts of genus present that were expected and genus expected
--samples_obs_count_type: average count of genus present across all samples by control type
--samples_obs_count_kit: average count of genus present across all samples by kit

**samples_obs_abund: percent abundance of zymo/ac samples

**samples_obs_abund_vari: percent abundance of by control, by kit, percent differences of kit by control

#1. Load required libraries
```{r}
library("biom")
library(data.table)
library(ggplot2)
```

#2. Taxonomy Reference - Expected frequencies
```{r}
#Read in controls taxonomy file, grouped by Genus
control_expect <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Controls\\Taxonomy\\Taxonomy_Genus.txt",sep="\t",header=TRUE)

#Create list of expected genus; add as the row names for control_expect
control_select <- subset(control_expect, !(MSA1000=='NA')|!(MSA1001=='NA')|!(MSA1002=='NA')|!(MSA1003=='NA')|!(Zymo.Seq=='NA'))
genus_expect <- unique(control_select[,"Genus"])
rownames(control_expect) <- unique(control_expect[,"Genus"])

#Create Control Composition Database, removing rows of species not in extraction control samples, specific to Seq control type
##msa's
control_m0 <- subset(control_expect, !(MSA1000=='NA'))
control_m0 <- control_m0[,c(1:10,12)]

control_m1 <- subset(control_expect, !(MSA1001=='NA'))
control_m1 <- control_m1[,c(1:10,13)]

control_m2 <- subset(control_expect, !(MSA1002=='NA'))
control_m2 <- control_m2[,c(1:10,14)]

control_m3 <- subset(control_expect, !(MSA1003=='NA'))
control_m3 <- control_m3[,c(1:10,15)]

#Zymo Sequencing Control
control_zymoseq <-subset(control_expect, !(Zymo.Seq=='NA'))
control_zymoseq <- control_zymoseq[,c(1:10,17)]
```

#3. Filtering of samples, control variable information
```{r}
#Read in database with variable information
samples_variables_complete <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Analysis\\ExpA_Run1&2\\Manifests\\sampledata.txt",sep="\t",header=TRUE)

#Read in database of failed samples
samples_failed <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Analysis\\ExpA_Run1&2\\TaxaCounts\\samplesfailed_q2.txt",sep="\t")

#Remove failed samples from info database
samples_filtered <-subset(samples_variables_complete, !(NephID %in% samples_failed$V1))

#Verify filtering - numbers should match
nrow(samples_variables_complete) - nrow(samples_failed)
nrow(samples_filtered)

#Only include Ext Controls
samples_variables_f <- subset(samples_filtered,Type=='MSA-1000' |Type=='MSA-1001' |Type=='MSA-1002' |Type=='MSA-1003' | Type=='Zymo.Seq.200'| Type=='Zymo.Seq.2000')

#Add sample ID's as row names
row.names(samples_variables_f) <- samples_variables_f$NephID
```

#4. OTU Counts by sample
Read in OTU information, and determine total counts for all samples
```{r}
#Read in OTU database at Genus level
samples_otu_complete <- read.csv("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Analysis\\ExpA_Run1&2\\TaxaCounts\\qiime2_taxa_counts_genus_silva.csv", row.names = 1)

#Create a list of all samples and all genus present in at least one sample
sample_list_complete <- colnames(samples_otu_complete)
genus_found <- rownames(samples_otu_complete)

#For each sample, determine the total read counts
for (a in sample_list_complete){
 #Set sum to 0
 sum=0
 
 #For each genus
 for (b in genus_found){
  #Add the number of counts together
  sum <- samples_otu_complete[b,a] + sum
 }
 
 #When complete, push the sum to a new "Total_Reads" row
 samples_otu_complete["Total_Reads",a] <- sum
}
```

#5. OTU Counts by Control Type
Determine genus count for each sample, average across control type
```{r}
#Create control sample list, control type list
sample_list_f <- unique(samples_variables_f$NephID)
control_types <- c("MSA-1000", "MSA-1001", "MSA-1002", "MSA-1003","Zymo.Seq.200","Zymo.Seq.2000")

#Create a missing list for non-annotated Genus
missing_anno <- list()

##For each sample in filtered list
for (a in sample_list_f){
 
 #Set count to 0
 count_present = 0
 count_absent = 0
 
 #Create genus list of expected controls for specific control types
 if(samples_variables_f[a,"Type"]=="MSA-1000"){
  genus_temp <- unique(control_m0$Genus)
  } else if(samples_variables_f[a,"Type"]=="MSA-1001"){
   genus_temp <- unique(control_m1$Genus)
  } else if(samples_variables_f[a,"Type"]=="MSA-1002"){
   genus_temp <- unique(control_m2$Genus)
  } else if(samples_variables_f[a,"Type"]=="MSA-1003"){
   genus_temp <- unique(control_m3$Genus)
  } else if(samples_variables_f[a,"Type"]=="Zymo.Seq.200"){
   genus_temp <- unique(control_zymoseq$Genus)
  } else if(samples_variables_f[a,"Type"]=="Zymo.Seq.2000"){
   genus_temp <- unique(control_zymoseq$Genus)
  } else {genus_temp <- data.frame()}
 
 #Convert to character
 genus_temp <- as.character(genus_temp)
 
 #For each genus is generated list, determine number of reads
 for (b in genus_temp){
  value <- samples_otu_complete[b,a]

  #If value is NA (meaning that the genus was not found in sample), add to the absent count
  #If reads are >0 then count_presence increases by 1
  #If reads are =0 then count_absence increases by 1 

  if(is.na(value) | is.null(value)) {
   missing_anno[[b]] <- "No Annotations"
   count_absent = count_absent +1
   next
  } else if (value>0) {
   count_present = count_present +1
  } else if (value==0){
   count_absent = count_absent +1
  } else{
   next
   }
 }
 
 #Add final count and length of the list to samples_variables_f dataframe
 samples_variables_f[a,"GenusObs"] <- count_present
 samples_variables_f[a,"GenusAbs"] <- count_absent
 samples_variables_f[a,"GenusExp"] <- length(genus_temp)
}
###Testing
write.csv(samples_variables_f,"temp1.csv")

#List of Genus that no samples contained
missing_anno

#New Database to store final Genus information
samples_obs_count_type <- data.frame()

#For each of the control types
for (b in control_types){
 
 #Reset counts
 genus=0
 count=0
 
 #For each of the samples in the sample list
 for(a in sample_list_f){
  
  #Create control type groups
  valid <- samples_variables_f[a,"Type"]
  
  #If the current samples control type matches the control type of the iteration
  if(valid==b){
   
   #Count the number of genus in the sample
   genus <- genus + samples_variables_f[a,"GenusObs"]
   count <- count + 1
  } else{
    next
  }
  #Find number of genus expected
  genus_expect_count <- samples_variables_f[a,"GenusExp"]
 }
 
 #Find the average of the genus and totals
 genus_average <- genus/count

 #Push the totals to the new database 
 samples_obs_count_type[1,b] <- genus_average
 samples_obs_count_type[2,b] <- genus_expect_count

 #Add Row names
 row.names(samples_obs_count_type) <- c("Genus Count", "Expected")
}

###Testing
write.csv(samples_obs_count_type,"temp2.csv")
```

#6. OTU Percent - Analysis Genus OTU Abundance
```{r}
#Create new table
samples_obs_abund = data.frame()

#Determine which genus_expect to use
##Run through each sample
for (a in sample_list_f){
 
 #Check for each of the expected genus
 for (b in genus_expect){
  
  #Determine the reads for the genus, and the total reads for the sample
  reads <- samples_otu_complete[b,a]
  total <- samples_otu_complete["Total_Reads",a]

  #If value is NA (meaning that the genus was not found in sample), skip
  #If reads are >0 then add the read count to reads_total

  if(is.na(reads) | is.null(reads)) {
   next
  } else {
   percent_exp <- (reads/total)
   
   #Add percent to the samples_obs_abund database
   samples_obs_abund[b,a] <- percent_exp
  }
 }
}

#Output the table of all samples
write.csv(samples_obs_abund,"Data/seq_obs_abund.csv")
```

#7. OTU Percent - Analysis Genus OTU Abundance by Control Type
```{r}
#New Database to store final Genus information
samples_obs_abund_vari <- data.frame()

#For each of the control types
for (a in control_types){
 
  #For each Genus in list
  for (b in genus_expect){
   
   #Reset Counter
   reads <- 0
   total <- 0
   
   #Create a list of samples that are of the specified control type
   sample_list_temp<- subset(samples_variables_f, Type==a)
   sample_list_temp <- unique(sample_list_temp$NephID)
   
   #For each of the samples in the sample list
   for(c in sample_list_temp){
    
    #Add up the total percentages from samples_obs_abund [genus,sampleID]
    reads <- samples_otu_complete [b,c] + reads
    total <- samples_otu_complete["Total_Reads",c] + total
   }
   
   #average of genus level by taking total number of reads of that genus divded by total number of reads for samples included
   #Push value to new dataframe
   percent_obs <- reads / total
   samples_obs_abund_vari[b,a] <- percent_obs
  }
 }

#Output to text file
write.csv(samples_obs_abund_vari,"Data/seq_obs_abund_vari.csv")
```

#10. OTU Percent - Analysis Exp OTU Rel Abun % Diff by Control by Kit
```{r}
#Add expected frequencies to data table for the AC
for (a in genus_expect){
 value1 <- control_expect[a,10] #gram
 value2 <- control_expect[a,12]
 value3 <- control_expect[a,13]
 value4 <- control_expect[a,14]
 value5 <- control_expect[a,15]
 value6 <- control_expect[a,16]
 value7 <- control_expect[a,17]

 #Add the Gram Positive / Neg Value
 if(is.na(value1) | is.null(value1)){
 } else{
  samples_obs_abund_vari[a,"Gram"] <- value1
 }
 
 #Add the MSA0 expected values
 if(is.na(value2) | is.null(value2)){
 } else{
  samples_obs_abund_vari[a,"M0_Exp"] <- value2
 }
 
  #Add the MSA1 expected values
 if(is.na(value3) | is.null(value3)){
 } else{
  samples_obs_abund_vari[a,"M1_Exp"] <- value3
 }
 
  #Add the MSA2 expected values
 if(is.na(value4) | is.null(value4)){
 } else{
  samples_obs_abund_vari[a,"M2_Exp"] <- value4
 }
 
  #Add the MSA3 expected values
 if(is.na(value5) | is.null(value5)){
 } else{
  samples_obs_abund_vari[a,"M3_Exp"] <- value5
 }
 
  #Add the Zymo Seq 200 expected value
 if(is.na(value6) | is.null(value6)){
 } else{
  samples_obs_abund_vari[a,"ZymoSeq200_Exp"] <- value6
 }
 
  #Add the Zymo Seq 2000 expected value
 if(is.na(value7) | is.null(value7)){
 } else{
  samples_obs_abund_vari[a,"ZymoSeq2000_Exp"] <- value7
 }
}

#Output to text file
write.csv(samples_obs_abund_vari,"Data/seq_obs_abund_vari.csv")
***************************** STOP
#Start Counters at Rel Abundance for ZS2000 (8) M2(9) M3 (10) Z200 (11) M0 (12) M1(13)
b = 8
c = 9
d = 10
e = 11
f = 12
g = 13

for (c in c(8:13)) #For each expected genus
 for (a in genus_expect){

  #Check if the control is expected - Zymo Ext.
  value1 <- samples_obs_abund_vari[a,c]
  value2 <- samples_obs_abund_vari[a,19]
  
  #Calculate percent difference
  if(is.na(value1) | is.null(value1)| is.na(value2) | is.null(value2)){
   value <- "NA"
  } else{
   value <- (samples_obs_abund_vari[a,c] - samples_obs_abund_vari[a,19])/ samples_obs_abund_vari[a,19]
  } 
 
  #Check if the control is expected - AC
  value1 <- samples_obs_abund_vari[a,b]
  value2 <- samples_obs_abund_vari[a,20]
  
  #Calculate percent difference
  if(is.na(value1) | is.null(value1)| is.na(value2) | is.null(value2)){
   value2 <- "NA"
  } else{
   value2 <- (samples_obs_abund_vari[a,b] - samples_obs_abund_vari[a,20])/ samples_obs_abund_vari[a,20]
  } 
  
  #Create column names 
  col_name <- paste((colnames(samples_obs_abund_vari[b])),"- Df")
  col_name2 <- paste((colnames(samples_obs_abund_vari[c])),"- Df")
 
  #Pass values
  samples_obs_abund_vari[a,col_name]<- value2
  samples_obs_abund_vari[a,col_name2]<- value
 }
 #Increase counter by 2, every other col is AC or ZymoExt
 b = b+2
 c = c+2
}

#Output to text file
write.csv(samples_obs_abund_vari,"Data/samples_obs_abund_vari.csv")
```

#11. OTU Percent - Analysis Unexp OTU Rel Abun by Control by Kit
```{r}
#Create full list of genus present
genus_present <- rownames(samples_otu_complete)
genus_expect<-as.character(genus_expect)

#Create list of unexpected
genus_unexpect <- genus_present[which(!genus_present %in% genus_expect)]

#Remove Total_reads from list
genus_unexpect <- genus_unexpect[genus_unexpect !="Total_Reads"]

#Create new dataframe
samples_obs_abund_vari_unexp <- data.frame()

#For each of the kits types
for (a in kit_types){
 
  #For each Genus in list
  for (b in genus_unexpect){
   
   #Reset Counter
   reads <- 0
   total <- 0
   
   #Create a list of samples that are of the specified control type
   sample_list_temp<- subset(samples_variables_f, Kit==a)
   sample_list_temp <- unique(sample_list_temp$NephID)
   
   #For each of the samples in the sample list
   for(c in sample_list_temp){
    
    #Add total reads from samples_otu_complete [genus,sampleID]
    reads <- samples_otu_complete [b,c] + reads
    total <- samples_otu_complete["Total_Reads",c] + total
   }
   
   #Calc Rel abund (total number of reads / total number of reads)
   #Push value to new dataframe
   percent_obs <- reads / total
   samples_obs_abund_vari_unexp[b,a] <- percent_obs
  }
}
#Output to text file
write.csv(samples_obs_abund_vari_unexp,"Data/samples_obs_abund_vari_unexp.csv")

#For each of the kits types
for (a in kit_types){
 
 #For each of the control types
 for(b in control_types){
  #Create a list of samples that are of the specified control type
  sample_list_temp<- subset(samples_variables_f, Kit==a & Type==b)
  sample_list_temp <- unique(sample_list_temp$NephID)
  
  #For each of the expected genus
  for(c in genus_unexpect){
   
   #Reset Counter
   reads <- 0
   total <- 0

   #For each of the samples in the sample list
   for(d in sample_list_temp){
    
    #Add reads from samples_otu_complete [genus,sampleID]
    reads <- samples_otu_complete [c,d] + reads
    total <- samples_otu_complete["Total_Reads",d] + total
   }
   
   #Calc Rel abund (total number of reads / total number of reads)
   percent_obs <- (reads/total)
   
   #Create new header
   #Push value to new dataframe
   col_name <- paste(a,"-",b)
   samples_obs_abund_vari_unexp[c,col_name] <- percent_obs
  }
 }
}
#Output to text file
write.csv(samples_obs_abund_vari_unexp,"Data/samples_obs_abund_vari_unexp.csv")
```

#12A Plot graph - Actual to Expected - Intrakit (Artificial Colonies)
For each kit, show variability, and map
```{r}
#Create genus list for art colonies only
genus_expect <- rownames(control_artcol)
genus_expect

plot_list <- list()
i=1

for (a in kit_types){
 temp<-data.frame()
 count=1
 
###Create database
 temp2 <- subset(samples_variables_f,Kit==a)
 temp2 <- subset(temp2,Type=="AC")

 #Subset, and create a sample list
 sample_list <- unique(temp2$NephID)
 sample_list
 
 #For each Genus in list, determine the samples value
 for (b in genus_expect){
    
   #For each of the samples in the sample list
   for(c in sample_list){
    temp[count,"Genus"]<-b
    temp[count,"Obs_Per"]<- samples_obs_abund[b,c]
    temp[count,"Sample"]<-c
    temp[count,"Exp_Per"]<-control_expect[b,"Art.Col"]
    count=count+1
   }
 }
 graph_title <- paste ("Artificial Colony:Intrakit Variablity",a)
 file_name <- paste("Data/",a,".jpeg")
 
 #Plot Data
 graph_new <- ggplot(data=temp,mapping=aes(x=Exp_Per,y=Obs_Per,color=Sample))+
 geom_point()+ #Can add points to graph
 geom_smooth(method = "lm",se=FALSE)+ #Add a smooth line
 labs(title=graph_title, x="Expected Percentage", y= "Actual Percentage")+
 geom_text(mapping=aes(x=Exp_Per,y=Obs_Per,label=Genus),check_overlap = TRUE)
 
 plot_list[[i]] = graph_new
 i=i+1
}

#Save all plots
for (i in 1:5) {
    file_name = paste("Data/Graphs/Intrakit Variability_AC_", kit_types[i], ".tiff", sep="")
    tiff(file_name)
    print(plot_list[[i]])
    dev.off()
}
```

#12B Plot graph - Actual to Expected - Intrakit (Zymo.Ext)
For each kit, show variability, and map
```{r}
#Create genus list for art colonies only
genus_expect <- rownames(control_zymoext)
genus_expect

plot_list <- list()
i=1

for (a in kit_types){
 temp<-data.frame()
 count=1
 
###Create database
 temp2 <- subset(samples_variables_f,Kit==a)
 temp2 <- subset(temp2,Type=="Zymo.Ext")

 #Subset, and create a sample list
 sample_list <- unique(temp2$NephID)
 sample_list
 
 #For each Genus in list, determine the samples value
 for (b in genus_expect){
    
   #For each of the samples in the sample list
   for(c in sample_list){
    temp[count,"Genus"]<-b
    temp[count,"Obs_Per"]<- samples_obs_abund[b,c]
    temp[count,"Sample"]<-c
    temp[count,"Exp_Per"]<-control_expect[b,"Zymo.Ext"]
    count=count+1
   }
 }
 graph_title <- paste ("Zymo Extraction:Intrakit Variablity",a)
 file_name <- paste("Data/",a,".jpeg")
 
 #Plot Data
 graph_new <- ggplot(data=temp,mapping=aes(x=Exp_Per,y=Obs_Per,color=Sample))+
 geom_point()+ #Can add points to graph
 geom_smooth(method = "lm",se=FALSE)+ #Add a smooth line
 labs(title=graph_title, x="Expected Percentage", y= "Actual Percentage")+
 geom_text(mapping=aes(x=Exp_Per,y=Obs_Per,label=Genus),check_overlap = TRUE)
 
 plot_list[[i]] = graph_new
 i=i+1
}

#Save all plots
for (i in 1:5) {
    file_name = paste("Data/Graphs/Intrakit Variability_ZE_", kit_types[i], ".tiff", sep="")
    tiff(file_name)
    print(plot_list[[i]])
    dev.off()
}
```

#13. Plot graph - Actual to Expected - All kits
```{r}
###############
#Zymo Ext
###############
#Create dataframe
zymo.kits <- samples_obs_abund_vari[,c(19,9,11,13,15,17)]
zymo.kits <- subset(zymo.kits,!(is.na(Zymo_Exp)))

#Check
rownames(zymo.kits)
colnames(zymo.kits)

#Create dataframe
temp<-data.frame()
count=1

#For each of the Kits
for (a in c(2:6)){
 count2=1
 
 #for each genus
 for(b in 1:nrow(zymo.kits)){
  temp[count,"Exp_Per"]<- zymo.kits[count2,"Zymo_Exp"]
  temp[count,"Genus"] <- rownames(zymo.kits[b,])
  temp[count,"Obs_Per"] <- zymo.kits[b,a]
  temp[count,"Kit"] <- colnames(zymo.kits[a])
  count=count+1
  count2=count2+1
 }
}
head(temp)

#Plot graph
graph_new<- ggplot(data=temp,mapping=aes(x=Exp_Per,y=Obs_Per,color=Kit))+
 #geom_point()+ #Can add points to graph
 geom_smooth(method = "lm",se=FALSE)+ #Add a smooth line
 labs(title="Zymo.Extraction - Kit Comparison", x="Expected Percentage", y= "Actual Percentage")+
 geom_text(mapping=aes(x=Exp_Per,y=Obs_Per,label=Genus),check_overlap = TRUE)

file_name =("Data/Graphs/Kit Differences_ZE_.tiff")
tiff(file_name)
print(graph_new)
dev.off()

###############
#AC
###############
#Create dataframe
ac.kits <- samples_obs_abund_vari[,c(20,8,10,12,14,16)]
ac.kits <- subset(ac.kits,!(is.na(AC_Exp)))

#Check
rownames(ac.kits)
colnames(ac.kits)

#Create dataframe
temp<-data.frame()
count=1

#For each of the Kits
for (a in c(2:6)){
 count2=1
 
 #for each genus
 for(b in 1:nrow(zymo.kits)){
  temp[count,"Exp_Per"]<- ac.kits[count2,"AC_Exp"]
  temp[count,"Genus"] <- rownames(ac.kits[b,])
  temp[count,"Obs_Per"] <- ac.kits[b,a]
  temp[count,"Kit"] <- colnames(ac.kits[a])
  count=count+1
  count2=count2+1
 }
}
head(temp)

#Plot graph
graph_new<- ggplot(data=temp,mapping=aes(x=Exp_Per,y=Obs_Per,color=Kit))+
 #geom_point()+ #Can add points to graph
 geom_smooth(method = "lm",se=FALSE)+ #Add a smooth line
 labs(title="Artificial Colony - Kit Comparison", x="Expected Percentage", y= "Actual Percentage")+
 geom_text(mapping=aes(x=Exp_Per,y=Obs_Per,label=Genus),check_overlap = TRUE)

file_name =("Data/Graphs/Kit Differences_AC_.tiff")
tiff(file_name)
print(graph_new)
dev.off()
```